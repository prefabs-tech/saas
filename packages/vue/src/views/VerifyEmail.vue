<template>
  <Page :title="t('user.emailVerification.title')" class="auth verify-email">
    <Card>
      <p>{{ message }}</p>

      <template
        v-if="
          status === EMAIL_VERIFICATION.EMAIL_VERIFICATION_INVALID_TOKEN_ERROR
        "
        #footer
      >
        <ButtonElement
          :label="t('user.emailVerification.button.label')"
          @click="handleResend"
        />
      </template>
    </Card>

    <LoadingPage :loading="loading" />
  </Page>
</template>

<script lang="ts">
export default {
  name: "VerifyEmail",
};
</script>

<script setup lang="ts">
import { useI18n } from "@dzangolab/vue3-i18n";
import { ButtonElement, Card, LoadingPage, Page } from "@dzangolab/vue3-ui";
import { computed, onMounted, ref } from "vue";

import { EMAIL_VERIFICATION } from "../constant";
import { useTranslations, emitter } from "../index";
import useUserStore from "../store";
import { resendVerificationEmail, verifyEmail } from "../supertokens";

const messages = useTranslations();

const { t } = useI18n({ messages });

const { user } = useUserStore();

const loading = ref<boolean>(false);
const status = ref<string>();

const message = computed(() => {
  if (loading.value) {
    return t("user.emailVerification.messages.verifyingEmail");
  }

  let message = "";

  switch (status.value) {
    case EMAIL_VERIFICATION.OK:
      message = t("user.emailVerification.messages.verify.success");
      break;
    case EMAIL_VERIFICATION.EMAIL_ALREADY_VERIFIED:
      message = t("user.emailVerification.messages.resend.alreadyVerified");
      break;
    case EMAIL_VERIFICATION.EMAIL_VERIFICATION_INVALID_TOKEN_ERROR:
      message = t("user.emailVerification.messages.verify.invalidToken");
      break;
    default:
      message = t("user.emailVerification.messages.resend.error");
  }

  return message;
});

onMounted(() => {
  if (user) {
    loading.value = true;

    verifyEmail()
      .then((response) => {
        status.value = response.status;

        if (status.value === EMAIL_VERIFICATION.OK) {
          emitter.emit("notify", {
            text: t("user.emailVerification.messages.verify.success"),
            type: "success",
          });
        } else if (
          status.value === EMAIL_VERIFICATION.EMAIL_ALREADY_VERIFIED_ERROR
        ) {
          emitter.emit("notify", {
            text: t("user.emailVerification.messages.verify.alreadyVerified"),
            type: "info",
          });
        } else {
          emitter.emit("notify", {
            text: t("user.emailVerification.messages.verify.invalidToken"),
            type: "error",
          });
        }
      })
      .catch(() => {
        emitter.emit("notify", {
          text: t("user.emailVerification.messages.resend.error"),
          type: "error",
        });
      })
      .finally(() => {
        loading.value = false;
      });
  }
});

const handleResend = () => {
  resendVerificationEmail()
    .then((verificationStatus) => {
      if (verificationStatus === EMAIL_VERIFICATION.OK) {
        emitter.emit("notify", {
          text: t("user.emailVerification.messages.resend.success"),
          type: "success",
        });
      } else if (
        verificationStatus === EMAIL_VERIFICATION.EMAIL_ALREADY_VERIFIED_ERROR
      ) {
        emitter.emit("notify", {
          text: t("user.emailVerification.messages.resend.alreadyVerified"),
          type: "info",
        });
      }
    })
    .catch(() => {
      emitter.emit("notify", {
        text: t("user.emailVerification.messages.resend.error"),
        type: "error",
      });
    });
};
</script>
