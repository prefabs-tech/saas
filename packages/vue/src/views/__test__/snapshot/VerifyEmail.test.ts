import i18nPlugin, { useLocaleStore } from "@dzangolab/vue3-i18n";
import { mount } from "@vue/test-utils";
import { createPinia } from "pinia";
import { describe, it, expect } from "vitest";

import appConfig from "../../../components/__test__/config";
import VerifyEmail from "../../VerifyEmail.vue";

import type { VueWrapper } from "@vue/test-utils";

describe("VerifyEmail", () => {
  const pinia = createPinia();

  const { setLocale } = useLocaleStore(appConfig.slug);

  const locales = ["en", "fr"];

  for (const locale of locales) {
    setLocale(locale);

    const wrapper: VueWrapper = mount(VerifyEmail, {
      global: {
        plugins: [
          pinia,
          [
            i18nPlugin,
            {
              config: appConfig,
            },
          ],
        ],
      },
    });

    it("matches snapshot in " + locale, () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  }
});
