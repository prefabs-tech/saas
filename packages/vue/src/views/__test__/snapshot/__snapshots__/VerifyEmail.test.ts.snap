// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VerifyEmail > matches snapshot in en 1`] = `
<div
  class="page auth verify-email"
  data-centered="false"
>
  <div
    class="page-header"
  >
    <div
      class="page-title-wrapper"
    >
      <div
        class="title"
      >
        <h1>
          Email verification
        </h1>
        
        <!---->
        
      </div>
      
      <!---->
      
    </div>
    <div
      class="page-toolbar"
    >
      
      
    </div>
  </div>
  <div
    class="page-content"
  >
    
    <div
      class="card"
    >
      <!---->
      <div
        class="card-body"
      >
        
        <p>
          Something went wrong. Please try again
        </p>
        
      </div>
      <!---->
    </div>
    <!---->
    
    <!---->
  </div>
</div>
`;

exports[`VerifyEmail > matches snapshot in fr 1`] = `
<div
  class="page auth verify-email"
  data-centered="false"
>
  <div
    class="page-header"
  >
    <div
      class="page-title-wrapper"
    >
      <div
        class="title"
      >
        <h1>
          Email verification
        </h1>
        
        <!---->
        
      </div>
      
      <!---->
      
    </div>
    <div
      class="page-toolbar"
    >
      
      
    </div>
  </div>
  <div
    class="page-content"
  >
    
    <div
      class="card"
    >
      <!---->
      <div
        class="card-body"
      >
        
        <p>
          Something went wrong. Please try again
        </p>
        
      </div>
      <!---->
    </div>
    <!---->
    
    <!---->
  </div>
</div>
`;
